examples/step-a/step-a.cc

```diff
@@ -0,0 +1,1654 @@
+    bool        read_bool_attr(const std::string attr_name) const;
+    std::string read_string_attr(const std::string attr_name) const;
+    std::vector<std::vector<double>>
+         read_matrix_attr(const std::string attr_name) const;
+    void write_int_attr(const std::string attr_name, int value) const;
+    void write_double_attr(const std::string attr_name, double value) const;
+
+  private:
+    hid_t             hdf5Group;
+    const bool        rootGroup;
+    const H5Group *   parentGroup;
+    const std::string filename;
+    const std::string groupName;
+    unsigned int *    instancesCounter;
+  };
+
+  H5Group::H5Group(const std::string filename_, const FileMode filemode_)
+    : rootGroup(true)
+    , parentGroup(nullptr)
+    , filename(filename_)
+    , groupName("Not_a_group")
+  {
+    switch (filemode_)
+      {
+        case (FileMode::create):
+          instancesCounter  = new unsigned int;
+          *instancesCounter = 1;
+          hdf5Group =
+            H5Fcreate(filename.data(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+          break;
+        case (FileMode::open):
+          instancesCounter  = new unsigned int;
+          *instancesCounter = 1;
+          hdf5Group = H5Fopen(filename.data(), H5F_ACC_RDWR, H5P_DEFAULT);
+          break;
+        default:
+          throw std::runtime_error(
+            std::string("Not implemented at H5Group::H5Group()"));
+          break;
+      }
+  }
+
+  H5Group::H5Group(const H5Group &parentGroup_, const std::string groupName_)
+    : rootGroup(false)
+    , parentGroup(&parentGroup_)
+    , filename("Not_a_file")
+    , groupName(groupName_)
+  {
+    instancesCounter  = new unsigned int;
+    *instancesCounter = 1;
+    if (H5Lexists(parentGroup->hdf5Group, groupName.data(), H5P_DEFAULT) != 0)
+      {
+        hdf5Group =
+          H5Gopen2(parentGroup->hdf5Group, groupName.data(), H5P_DEFAULT);
+      }
+    else
+      {
+        hdf5Group = H5Gcreate2(parentGroup->hdf5Group,
+                               groupName.data(),
+                               H5P_DEFAULT,
+                               H5P_DEFAULT,
+                               H5P_DEFAULT);
+      }
+  }
+
+  H5Group::H5Group(const H5Group &other)
+    : hdf5Group(other.hdf5Group)
+    , rootGroup(other.rootGroup)
+    , parentGroup(other.parentGroup)
+    , filename(other.filename)
+    , groupName(other.groupName)
+    , instancesCounter(other.instancesCounter)
+  {
+    ++*instancesCounter;
+  }
+
+  H5Group::H5Group(H5Group && /*other*/)
+    : hdf5Group(0)
+    , rootGroup(false)
+    , parentGroup(nullptr)
+    , filename("Not_a_file")
+    , groupName("Not_a_group")
+    , instancesCounter(nullptr)
+  {
+    throw std::runtime_error(
+      std::string("Not implemented at move copy constructor H5Group"));
+  }
+
+  H5Group &H5Group::operator=(const H5Group & /*other*/)
+  {
+    throw std::runtime_error(
+      std::string("Not implemented at Copy Assignment Operator H5Group"));
```

Comment:
couldn't you just `=delete` this (and the next) operator?

Reply:
oh, and the previous one as well? if they can't be called at run time, maybe we should just disallow them at compile time as well.

Reply:
Done!